File: algorithm_data_structure/leetcode/455.assign-cookies.py
#
# @lc app=leetcode id=455 lang=python3
#
# [455] Assign Cookies
#

# @lc code=start
class Solution:
    def findContentChildren(self, g: List[int], s: List[int]) -> int:
        # Greedy algorithm

        # Sort to easily get the largest g and s.
        g.sort()
        s.sort()

        num_children = 0
        i = len(g) - 1
        j = len(s) - 1

        while i >= 0 and j >= 0:
            if s[j] >= g[i]:
                num_children += 1
                i = i - 1
                j = j - 1
            else:
                # Move to the next child, but still uses the largest cookie
                i = i - 1

        return num_children


# @lc code=end

File: algorithm_data_structure/leetcode/455.assign-cookies.cpp
/*
 * @lc app=leetcode id=455 lang=cpp
 *
 * [455] Assign Cookies
 */

// @lc code=start

#include <algorithm>
#include <vector>

class Solution
{
public:
    int findContentChildren(vector<int> &g, vector<int> &s)
    {
        // Greedy algorithm -> Start from the child with the minimal greed factor
        std::sort(g.begin(), g.end());
        std::sort(s.begin(), s.end());

        int child = 0;
        int cookie = 0;

        // Iterate over all children and cookies
        while (child < g.size() && cookie < s.size())
        {
            // This cookie can feed the current child.
            if (g[child] <= s[cookie])
            {
                ++child;
            }
            ++cookie;
        }

        return child;
    }
};
// @lc code=end
